### ISSUE TITLE
Kubeflow pipeline from elyra are exposing cerds

### ISSUE THREAD

**Issue by harshad16** [operate-first/support#155]
**Describe the bug**
we noticed that when we use elyra and kubeflow pipelines, the aws cerds and other information are being stored as env vars. which are being exposed on the kubeflow pipelines and tekton pipeline dashboard as the status show the configuration and environment variable value gets displayed.

 It should be suggested to the operate-first users to refrain from using the aws cerds for now on the elyra pipelines.
upstream issue is opened: https://github.com/elyra-ai/elyra/issues/1479
Related-to: https://github.com/aicoe-aiops/ocp-ci-analysis/issues/190

These can be seen at kubeflow pipelines as well: http://istio-ingressgateway-istio-system.apps.zero.massopen.cloud/_/pipeline/#/experiments

**To Reproduce**
Go to 'https://tekton.operate-first.cloud/#/namespaces/kubeflow/pipelineruns/lambda-ed231'
Click on status
Scroll down to step 1
See issues

**Expected behaviour**
The secrets are safe in openshift secrets and the configuration uses the secretref instead of direct environment variable value.

**Comment by tumido** [comment]
This has to be fixed upstream. I don't think there's anything we can do about it. We can warn users though.

/kind documentation
/sig user-experience

**Comment by pacospace** [comment]
> This has to be fixed upstream. I don't think there's anything we can do about it. We can warn users though.
> 
> /kind documentation
> /sig user-experience

What about creating a bucket with public credentials (playpen bucket) as of know (username: `operatefirst`, password: `operatefirstzero` that anyone can use to at least test pipelines, maybe with data can be public or also just trying to learn about pipelines?

**Comment by tumido** [comment]
we can't set the credentials, but we have one bucket like that already. It should be used by the data catalog by default but that's stuck on me fixing the data catalog. :smile:

**Comment by ptitzler** [comment]
An [enhancement that will address this issue](https://github.com/elyra-ai/elyra/pull/1529) should be available in the next release of Elyra.

TL;DR:
 - User creates a Kubernetes secret that contains the credentials
 - User references secret in runtime configuration
 - During pipeline execution the credentials are supplied to the pods executing the notebooks | scripts in an opaque way

**Comment by ptitzler** [comment]
The fix has been delivered and should be available in the `s2i-lab-elyra:v0.0.7` image. There are two additional steps users need to complete:
- create a Kubernetes secret that contains the S3 credentials in the user's namespace (this needs to be done manually)
- specify the secret name in the Kubeflow runtime configuration (this part is important; if no secret is specified Elyra behaves the same way it did in earlier releases)

![image](https://user-images.githubusercontent.com/13068832/115608809-9a763200-a29b-11eb-9c5f-e9bdf412ba03.png)

> Note that the credentials (uid/pwd) must still be specified in the runtime configuration! Elyra is not a Kubernetes native application (most users run it from a terminal) and therefore doesn't have access to Kubernetes objects.

When the pipeline is executed Kubeflow will automatically make the credentials from the secret available to the containers that are processing the notebooks/scripts.

Details are in the updated documentation at https://elyra.readthedocs.io/en/latest/user_guide/runtime-conf.html#cloud-object-storage-credentials-secret-cos-secret

@pacospace this impacts the tutorial/lab

**Comment by pacospace** [comment]
Thanks @ptitzler, I opened: https://github.com/thoth-station/elyra-aidevsecops-tutorial/issues/136

> The fix has been delivered and should be available in the `s2i-lab-elyra:v0.0.7` image. There are two additional steps users need to complete:
> 
> * create a Kubernetes secret that contains the S3 credentials in the user's namespace (this needs to be done manually)

As user of JH on operate first I might not have requested a namespace for myself, but rather working on JupyterHub only on my notebooks. 

With this solution every user now needs to have their own namespace if they want to submit a secure pipeline with Elyra. So it becomes a pre-requisite to use pipelines on Elyra. 

> * specify the secret name in the Kubeflow runtime configuration (this part is important; if no secret is specified Elyra behaves the same way it did in earlier releases)

Perfect!

> 
> ![image](https://user-images.githubusercontent.com/13068832/115608809-9a763200-a29b-11eb-9c5f-e9bdf412ba03.png)
> 
> > Note that the credentials (uid/pwd) must still be specified in the runtime configuration! Elyra is not a Kubernetes native application (most users run it from a terminal) and therefore doesn't have access to Kubernetes objects.
> 
Why not having a tab that make one case or the other mandatory? Use secret then make it mandatory and the other parameters no more mandatory.

> When the pipeline is executed Kubeflow will automatically make the credentials from the secret available to the containers that are processing the notebooks/scripts.

> 
> Details are in the updated documentation at https://elyra.readthedocs.io/en/latest/user_guide/runtime-conf.html#cloud-object-storage-credentials-secret-cos-secret
> 
> @pacospace this impacts the tutorial/lab

**Comment by pacospace** [comment]
> The fix has been delivered and should be available in the `s2i-lab-elyra:v0.0.7` image. There are two additional steps users need to complete:
> 
> * create a Kubernetes secret that contains the S3 credentials in the user's namespace (this needs to be done manually)
> * specify the secret name in the Kubeflow runtime configuration (this part is important; if no secret is specified Elyra behaves the same way it did in earlier releases)
> 
> ![image](https://user-images.githubusercontent.com/13068832/115608809-9a763200-a29b-11eb-9c5f-e9bdf412ba03.png)
> 
> > Note that the credentials (uid/pwd) must still be specified in the runtime configuration! Elyra is not a Kubernetes native application (most users run it from a terminal) and therefore doesn't have access to Kubernetes objects.
> 
> When the pipeline is executed Kubeflow will automatically make the credentials from the secret available to the containers that are processing the notebooks/scripts.
> 
> Details are in the updated documentation at https://elyra.readthedocs.io/en/latest/user_guide/runtime-conf.html#cloud-object-storage-credentials-secret-cos-secret
> 
> @pacospace this impacts the tutorial/lab

I'm running elyra v0.0.7 and I don't see that actually:

![Screenshot from 2021-04-21 22-32-36](https://user-images.githubusercontent.com/27498679/115617383-a12f9400-a2f1-11eb-8931-8637ec7cf7fd.png)

**Comment by ptitzler** [comment]
> I'm running elyra v0.0.7 and I don't see that 

Aww, I'm sorry. My apologies. I've mixed up things. [The PR is included in the next release](https://github.com/elyra-ai/elyra/pull/1529), which isn't available yet. We just published 2.2.2 but didn't retrofit this enhancement.

**Comment by ptitzler** [comment]
Nevertheless, to answer your questions:

> With this solution every user now needs to have their own namespace if they want to submit a secure pipeline with Elyra. So it becomes a pre-requisite to use pipelines on Elyra.

Yes, in shared environments it does.

> Why not having a tab that make one case or the other mandatory? Use secret then make it mandatory and the other parameters no more mandatory.

I expected the question (: We cannot do that because Elyra does not have any Kubernetes awareness and therefore cannot access or manipulate any Kubernetes objects. The issue is that the credentials are required when Elyra prepares the pipeline for execution. As part of that preparation it generates archives and uploads them to the specified cloud storage bucket. Since Elyra doesn't have access to Kubernetes secrets, there's unfortunately still the need to store the credentials in the runtime configuration.

**Comment by pacospace** [comment]
> Nevertheless, to answer your questions:
> 
> > With this solution every user now needs to have their own namespace if they want to submit a secure pipeline with Elyra. So it becomes a pre-requisite to use pipelines on Elyra.
> 
> Yes, in shared environments it does.
> 
> > Why not having a tab that make one case or the other mandatory? Use secret then make it mandatory and the other parameters no more mandatory.
> 
> I expected the question (: We cannot do that because Elyra does not have any Kubernetes awareness and therefore cannot access or manipulate any Kubernetes objects. The issue is that the credentials are required when Elyra prepares the pipeline for execution. As part of that preparation it generates archives and uploads them to the specified cloud storage bucket. Since Elyra doesn't have access to Kubernetes secrets, there's unfortunately still the need to store the credentials in the runtime configuration.

Thanks @ptitzler I ll make sure we describe this in the tutorial!! And we can see where to place these secrets of the user, if each user can see only their own secrets or if we add the pre-requisite in the tutorial for secure Elyra pipeline, cc @harshad16 @tumido @HumairAK @4n4nd

**Comment by ptitzler** [comment]
Please correct me if I'm wrong, but the Kubeflow setup on operate first is a single-user setup? When I connected to the web console the namespace appeared to be a static value that is not tied to my user id. If it is a single user deployment, then the pipeline pods are running in the `kubeflow` namespace. Only in multi-user mode are pods executed in the user's namespace.

**Comment by ptitzler** [comment]
We are going to include this functionality in the 2.2.3 release and will make it available in `s2i-lab-elyra:v0.0.8` fairly soon (🤞 ).

Quick update on my previous comment. I tested this in my environment (KF 1.2 & 1.3) and it appears that the pods are running in a user namespace after all (instead of `kubeflow`) if you created one as part of the setup. (Later releases of KFP/1.2+ started to require a user namespace, whereas earlier releases did not). I haven't had a chance to try this in the MOC/operate-first env yet.

**Comment by pacospace** [comment]
> We are going to include this functionality in the 2.2.3 release and will make it available in `s2i-lab-elyra:v0.0.8` fairly soon ( ).
> 
> Quick update on my previous comment. I tested this in my environment (KF 1.2 & 1.3) and it appears that the pods are running in a user namespace after all (instead of `kubeflow`) if you created one as part of the setup. (Later releases of KFP/1.2+ started to require a user namespace, whereas earlier releases did not). I haven't had a chance to try this in the MOC/operate-first env yet.

Hey @ptitzler,

running the new image v0.0.8 with Elyra 2.2.4 does not show the new feature you mentioned. You can see the screenshot here:

![Screenshot from 2021-05-12 16-24-19](https://user-images.githubusercontent.com/27498679/117992285-1addff00-b33f-11eb-905e-9a41cd8c432e.png)

**Comment by ptitzler** [comment]
It's [this setting](https://elyra.readthedocs.io/en/v2.2.4/user_guide/runtime-conf.html#cloud-object-storage-credentials-secret-cos-secret) in your screen shot: ![image](https://user-images.githubusercontent.com/13068832/117993047-3e2f8c80-b2f4-11eb-8adf-e49167f7dc64.png)

**Comment by pacospace** [comment]
> It's [this setting](https://elyra.readthedocs.io/en/v2.2.4/user_guide/runtime-conf.html#cloud-object-storage-credentials-secret-cos-secret) in your screen shot: ![image](https://user-images.githubusercontent.com/13068832/117993047-3e2f8c80-b2f4-11eb-8adf-e49167f7dc64.png)

Sorry I didn't catch that, thanks @ptitzler!!!

**Comment by ptitzler** [comment]
No worries. I guess its purpose can be misinterpreted as a password field because the input is hidden.

**Comment by harshad16** [comment]
Thanks all for working on this and fixing this issue.
